Add basic client-server application

gmkv opened the pull request:
The purpose of this simple client-server application is to demonstrate the
connection between two applications via TCP sockets.

mgenov commented on client-server-hello/client/client.go:
```diff
@@ -0,0 +1,51 @@
+	Address     *net.TCPAddr
+	connection  *net.TCPConn
+	network     string
+	IsConnected bool
```
used only internally so it could be with internal visibility.

mgenov commented on client-server-hello/client/client.go:
```diff
@@ -0,0 +1,51 @@
+
+type Client struct {
+	Address     *net.TCPAddr
+	connection  *net.TCPConn
```
Just conn as it's shorter and follows Go naming convention.

mgenov commented on client-server-hello/client/client.go:
```diff
@@ -0,0 +1,51 @@
+		nil
+}
+
+func (c *Client) Connect() error {
```
The client should be able to disconnect too.

defer client.Close() 
or
defer client.Disconnect()

mgenov commented on client-server-hello/client/client_test.go:
```diff
@@ -0,0 +1,143 @@
+		t.Errorf("client failed to connect to open port; error: %v", err)
+	}
+
+	// err = c.Connect()
```
Commented code need to be removed.

mgenov commented on client-server-hello/client/client_test.go:
```diff
@@ -0,0 +1,143 @@
+}
+
+func TestConnectToOpenPort(t *testing.T) {
+	addr := ":8085"
```
":0" for using of random free port. CI server may use this port which may cause test to fail.

mgenov commented on client-server-hello/client/client_test.go:
```diff
@@ -0,0 +1,143 @@
+
+}
+
+func TestErrorWhenConnectingAgain(t *testing.T) {
```
Seems incomplete.

mgenov commented on client-server-hello/client/client_test.go:
```diff
@@ -0,0 +1,143 @@
+
+}
+
+func TestCanReceive(t *testing.T) {
```
TestReceiveWhatWasSendFromServer

mgenov commented on client-server-hello/cmd/server-app/main.go:
```diff
@@ -0,0 +1,31 @@
+		panic(err)
+	}
+
+	err = s.Connect()
```
Server may start and connect sounds misleading in this context.

s.Start()

mgenov commented on client-server-hello/m.go:
```diff
@@ -0,0 +1,14 @@
+)
+
+func main() {
+	s, _ := server.New(":46532")
```
The API usage looks really strange.
s, _ := server.New(":46532")
s.Start()

mgenov commented on client-server-hello/m.go:
```diff
@@ -0,0 +1,14 @@
+	s, _ := server.New(":46532")
+	s.Connect()
+	s.StartAcceptingWithGreeting("")
+	<-time.After(10 * time.Minute)
```
It is better signals to be used instead for graceful termination.

https://golang.org/pkg/os/signal/
https://gobyexample.com/signals

mgenov commented on client-server-hello/server/server.go:
```diff
@@ -0,0 +1,101 @@
+
+//
+// if message is "", then write "hello + current time"
+func (s *Server) StartAcceptingWithGreeting(message string) error {
```
If the goal is StartAcceptingWithFunc to bee re-usable call you have to use it here too.

gmkv commented:
I have made all the fields of Client and Server structs unexported, as they are not used externally, 
and have also renamed their functions to clarify their behaviour.
I am reusing AcceptWithFunc to work with any connection handler function as I have originally intended.
Changed the main.go apps to await signals from the OS for proper process termination.

> ":0" for using of random free port. CI server may use this port which may cause test to fail.

This worked on the client unit tests, but I could not get this to work on the tests for the server - got a ":0 connection refused" error.

mgenov commented on client-server-hello/client/client_test.go:
```diff
@@ -0,0 +1,238 @@
+
+func TestClientInstantiatesCorrectly(t *testing.T) {
+	addr := ":0"
+	c, err := client.New(addr)
```
client.New(":0") to make it clear.

mgenov commented on client-server-hello/client/client_test.go:
```diff
@@ -0,0 +1,238 @@
+
+	// "" gets parsed as ":0"
+	addr = ""
+	c, err = client.New(addr)
```
client.New("")

Note that single test case for bad address is enough as you are now testing from here the ResolveTCPAddr implementation which should be well tested in the library.

mgenov commented on client-server-hello/client/client_test.go:
```diff
@@ -0,0 +1,238 @@
+		t.Errorf("failed to resolve address; error: %v", err)
+	}
+
+	testServer, err := net.ListenTCP("tcp", tcpAddr)
```
It is better TestServer type to be defined as it will simplify tests.

```golang
// Note that test server is created by passing the actual time
// which to be returned from it. This will let the client verify 
// the received result.
s, err := tcptest.NewServer(":0", time.Date(2009, time.November, 10, 23, 0, 0, 0, time.UTC)))
if err != nil {
 
}
defer s.Close()
```

Typical example of such package is: https://golang.org/pkg/net/http/httptest/

mgenov commented:
> I discovered a problem with the way the Server's AcceptWithFunc works.
Server's GreetFunc returns a closure, which AcceptWithFunc takes, as I wanted to make it reusable. However, since the returned function is a closure and it's supposed to get the current time, it only receives the current time. So, when many clients connect, they get identical time (which is the exact moment in time that AcceptWithFunc got its handler delegated.

Check the example that I provided with the test server. If you have to deal with callback in your test it should be easy as the testing callback may be created with specified clock or time.


```golang
type Clock interface {
  func Now() time.Time
}

type dummyClock struct {
  t time.Time
}
func NewDummyClock(t time.Time) dummyClock {
  return dummyClock{t}
}
func (d dummyClock) Now() time.Time {
   return d.t
}
```

So the GreetFunc is going to be created with:

```golang
greetFunc := server.NewGreetFunc(servertest.NewDummyClock(time.Date(2009, time.November, 10, 23, 0, 0, 0, time.UTC)))
```

e.g the time will be extracted from provider interface which in this case is named Clock.

gmkv commented:
Created a test server package for the client unit tests, and implemented the Clock interface for the server tests and the server package itself.

mgenov commented:
Ok, will take a look into it later today.

On Thu, Nov 1, 2018, 14:16 Georgi <notifications@github.com> wrote:

> Created a test server package for the client unit tests, and implemented
> the Clock interface for the server tests and the server package itself.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/clouway-incubator/georgi_minkov_go_networking/pull/1#issuecomment-435022894>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAMxR0t2b77NUYqrDVYzDviVEksx4cB4ks5uquYWgaJpZM4XwRk_>
> .
>
